File: 0684-redundant-connection/0684-redundant-connection.java
class Solution {
    
    public int[] findRedundantConnection(int[][] edges) {
        int[] arr = new int[2];
        final DSU dsu = new DSU(edges.length);

        for(int i = 0; i < edges.length; i++){
            int[] edge = edges[i];
            if (dsu.connected(edge[0], edge[1])){
                return edge;
            }
            dsu.union(edge[0],edge[1]);
        }
        return arr;
    }
}

public class DSU {
            private int[] parent;
            private int[] rank;

            public DSU(int n) {
                parent = new int[n+1];
                rank = new int[n+1];
                for (int i = 0; i < n; i++) {
                    parent[i] = i; // Initially, every element is its own parent
                    rank[i] = 0;   // Initial rank is 0
                }
            }

            // Find with Path Compression
            public int find(int i) {
                if (parent[i] == i) {
                    return i;
                }
                // Recursively find the root and make it the parent of i
                return parent[i] = find(parent[i]);
            }

            // Union by Rank
            public void union(int i, int j) {
                int rootI = find(i);
                int rootJ = find(j);

                if (rootI != rootJ) {
                    // Attach smaller rank tree under root of higher rank tree
                    if (rank[rootI] < rank[rootJ]) {
                        parent[rootI] = rootJ;
                    } else if (rank[rootI] > rank[rootJ]) {
                        parent[rootJ] = rootI;
                    } else {
                        parent[rootI] = rootJ;
                        rank[rootJ]++;
                    }
                }
            }

            // Check if two elements are in the same set
            public boolean connected(int i, int j) {
                return find(i) == find(j);
            }
        }

File: 0684-redundant-connection/0684-redundant-connection.py
class Solution(object):
    def findRedundantConnection(self, edges):
        """
        :type edges: List[List[int]]
        :rtype: List[int]
        """
        dsu = DSU(len(edges))
        for edge in edges:
            if dsu.connected(edge[0], edge[1]):
                return edge
            dsu.union(edge[0], edge[1])
        return []

class DSU:
    def __init__(self, n):
        self.parent = list(range(n + 1))
        self.rank = [0] * (n + 1)

    def find(self, i):
        if self.parent[i] != i:
            self.parent[i] = self.find(self.parent[i])  # Path compression
        return self.parent[i]

    def union(self, i, j):
        root_i, root_j = self.find(i), self.find(j)
        if root_i == root_j:
            return
        if self.rank[root_i] < self.rank[root_j]:
            self.parent[root_i] = root_j
        elif self.rank[root_i] > self.rank[root_j]:
            self.parent[root_j] = root_i
        else:
            self.parent[root_i] = root_j
            self.rank[root_j] += 1

    def connected(self, i, j):
        return self.find(i) == self.find(j)